sdk/core/azure-core/tests/azure_core_asynctests/test_tracing_decorator_async.py

```diff
@@ -0,0 +1,179 @@
+        self.pipeline = Pipeline(self.transport, policies=policies)
+
+        self.expected_response = mock.Mock(spec=PipelineResponse)
+        self.assert_current_span = assert_current_span
+
+    def verify_request(self, request):
+        if self.assert_current_span:
+            assert execution_context.get_current_span() is not None
+        return self.expected_response
+
+    @distributed_trace_async
+    async def make_request(self, numb_times, **kwargs):
+        time.sleep(0.001)
+        if numb_times < 1:
+            return None
+        response = self.pipeline.run(self.request, **kwargs)
+        await self.get_foo(merge_span=True)
+        kwargs['merge_span'] = True
+        await self.make_request(numb_times - 1, **kwargs)
+        return response
+
+    @distributed_trace_async
+    async def merge_span_method(self):
+        return await self.get_foo(merge_span=True)
+
+    @distributed_trace_async
+    async def no_merge_span_method(self):
+        return await self.get_foo()
+
+    @distributed_trace_async
+    async def get_foo(self):
+        time.sleep(0.001)
+        return 5
+
+    @distributed_trace_async(name_of_span="different name")
+    async def check_name_is_different(self):
+        time.sleep(0.001)
+
+    @distributed_trace_async(tracing_attributes={'foo': 'bar'})
+    async def tracing_attr(self):
+        time.sleep(0.001)
+
+    @distributed_trace_async
+    async def raising_exception(self):
+        raise ValueError("Something went horribly wrong here")
+
+
+@pytest.mark.usefixtures("fake_span")
+class TestAsyncDecorator(object):
+
+    @pytest.mark.asyncio
+    async def test_decorator_tracing_attr(self):
+        with FakeSpan(name="parent") as parent:
+            client = MockClient()
+            await client.tracing_attr()
+
+        assert len(parent.children) == 2
+        assert parent.children[0].name == "MockClient.__init__"
+        assert parent.children[1].name == "MockClient.tracing_attr"
+        assert parent.children[1].attributes == {'foo': 'bar'}
+
+
+    @pytest.mark.asyncio
+    async def test_decorator_has_different_name(self):
+        with FakeSpan(name="parent") as parent:
+            client = MockClient()
+            await client.check_name_is_different()
+        assert len(parent.children) == 2
+        assert parent.children[0].name == "MockClient.__init__"
+        assert parent.children[1].name == "different name"
+
+
+    @pytest.mark.asyncio
+    async def test_used(self):
+        with FakeSpan(name="parent") as parent:
+            client = MockClient(policies=[])
+            await client.get_foo(parent_span=parent)
+            await client.get_foo()
+
+            assert len(parent.children) == 3
+            assert parent.children[0].name == "MockClient.__init__"
+            assert not parent.children[0].children
+            assert parent.children[1].name == "MockClient.get_foo"
+            assert not parent.children[1].children
```

Comment:
no assertions for third child?

Reply:
he's grounded :)
we could, but that's not the point of this test, doesn't hurt to add it though

Reply:
fixed